### ISSUE TITLE
Policy for handling app specific FPs

### ISSUE THREAD

**Issue by lifeforms** [SpiderLabs/owasp-modsecurity-crs#527]
Before we go chasing large numbers of FPs, we have to think up a unified policy to handle application specific FPs.

As for timing, there's something to doing a "tick-tock" cadence of improving detection before RC1 and then doing a grind of FP reduction before RC2. I'm happy to look at every FP report in detail before RC2, but our time is limited. So I want to spend most effort on the big picture and the least time on workarounds.

I would be happy to commit to the goal of having zero FP for common applications like Drupal, though it's quite a commitment. CMS editor interfaces are very prone to many rule alerts, as people submit various HTML and content. We might not achieve this in a month. But we could get pretty far. However, if we commit to this goal, we need to do it in a very efficient and scalable way. Finetuning CRS rules themselves to exclude parameters for these applications (e.g. with chains) would in my opinion be laborious, and add lots of complexity to the rules which is hard to untangle later.

For instance, in #517 an overbroad pmf rule causes FP in `ARGS:ajax_html_ids[]` in the Drupal CMS editor interface. We could battle this specific FP in various ways. In the long term though, we will get hundreds of FP reports like this - we know how life was with CRS2 for a small installation, but now we are committing to suppressing them on a global scale.

To completely make CMS editing zero-FP, I think it is a viable long term strategy to start curating an application-specific rule file (ie 9xx-DRUPAL.conf) which would include app specific exclusions on selected API endpoints and ARGS. We could start doing the same for WordPress.

Keeping app specific exclusion files would be especially attractive if common parameters like `ajax_html_ids[]` get blocked by multiple rules, which based on my Drupal experience with CRS2 I am positive will happen, especially in higher paranoia levels. App specific exclusion rules can be maintained locally but do a broad sweep across many rules and paranoia levels. Especially when running on PL2+ there might be 5-15 attack rules involved with certain well-known application parameters like `ajax_html_ids[]`.

Separate exclusions also prevent base rule complexity. It's also harder for exclusion rules to introduce bugs or bypasses in unrelated APIs. If we really want to carve out a very specific application FP, you'll need multiple chains to check the URI, ARGS etc. otherwise you will open up `ajax_html_ids[]` for all kinds of unrelated APIs. That's a lot of chains for a single FP type. Now, `ajax_html_ids[]` is a very specific arg name and might not cause bypass in any other apps, but what about an FP in a more common `id` or `name` parameter which we'll surely get soon?

Finally it's easier to create exclusion rules. It's basically like running production and fixing FPs locally, which we'll have to do anyway on our projects; we'd just share those exclusions with the world. The barrier to maintain them will therefore also be lower for a CRS user when compared to hacking on a CRS attack rule itself. So exclusions may invite more contributions.

Also, people not desiring app specific exclusions might remove the exclusions by tag, which is impossible if we patch the attack rules in-place.

So it seems to me a productive activity to say, okay, we explicitly decide to be able to do CMS editing, wiki editing, in the default configuration with near-zero FP, then our best option in the general case is probably to start some app specific exclusion rule sets.

So my first proposal would be:

- RC1 focus on adding detection (which we have done)
- RC2 focus on FP (which considered the new issues created seems to be the case)
- Consider application specific exclusion files (eg. Drupal, WordPress, Mediawiki) rather than patching up rules

Comment, rate, and subscribe!

**Comment by csanders-git** [comment]
Honestly, there are some things i'd argue not to fix. For instance when you modify PHP in wordpress - it sends back a ton of PHP... and it is literally a PHP injection. I think fixing this would be bad, perhaps we could make separate optional rules for wordpress and drupal  but running those into the mainline doesn't really seem to be like something that offers an immense amount of support. previously I think the burden was that it would affect multiple apps (google analytics and such)

**Comment by dune73** [comment]
I like the idea of separate optional rule files for specific apps which would disable selected rules from the main CRS rule files. That is the right long term perspective. The maintainability and simplicity of the main CRS rules files is a big plus in my eyes.

The question is if it is the right moment to start it and if it it would not be better to add anti-fp chains into the main rule files for some time and consolidate in 3.1 or 3.2.

Walter: Could you give a syntax example? Perhaps for #517.

Things to consider: If we go out and state we try to avoid most if not all FPs in default install in 3.0 we need a way to handle the FP issues in an efficient way (or people will give up and the reputation of the project suffers). If we bring a new mechanism in RC2 we risk a release with instabilities.

Ahoj,

Christian

**Comment by csanders-git** [comment]
I think it will depend on the [amount of] feedback we get.

**Comment by lifeforms** [comment]
@dune73 Thinking out loud, I think people could activate the mode like this - perhaps in their crs-setup.conf, but if they're paranoid, perhaps just in the virtualhosts which are running Drupal specifically!

```
SecAction id:123456,phase:1,nolog,pass,t:none,setvar:tx.crs_tuning_drupal=1
```

I don't know if 'tuning' is the best term, but a blanket term might be nice, we might do exclusions as well as virtual patching in the future.

The Drupal file could go like this (pseudocode)...

```
SecRule &TX:crs_tuning_drupal|TX:crs_tuning_drupal "@eq 0" \
	"id:960000,\
	phase:1,\
	t:none,\
	nolog,\
	pass,\
	skipAfter:skipAfter:END_TUNING_DRUPAL"

# "edit-field-links-und-0-field-remove-item"
SecRule REQUEST_URI "/file/ajax/(submitted|upload)/" \
	"id:960010,\
	phase:1,\
	t:none,\
	nolog,\
	pass,\
	chain"
	SecRule ARGS:ajax_html_ids[] "^[a-z0-9\-]+$" \
		"t:none,\
		ctl:ruleRemoveTargetById=932120;ARGS:ajax_html_ids[]"

SecMarker END_TUNING_DRUPAL
```

When making these exclusions (960010-960020) we should be as specific as possible.

As for the rules to remove: here we have only one rule ID based on your report, but it could be expanded to a range or a tag. For a password field we could exclude it from the complete CRS (maybe `ctl:ruleRemoveTargetByTag=attack` which will match all rules having *some* attack tag).

As for the conditions to activate the exclusion: we should also validate the input if possible. This prevents from opening a hole in a separate API. In this example, I assume that `ARGS:ajax_html_ids[]` parameter is always alphanumeric and dashes. This should relieve us of a lot of worry associated with introducing false negatives.

It seems maybe a bit drastic to introduce this mechanism in 3.0. Maybe we shouldn't do it - at the other hand, it seems much better to me than doing hotfixes in main rules in RC2 - that's also a possibility for bugs, and we have to spend future time reverting them plus set up a future proof construct anyway :(

Comments welcome!

**Comment by dune73** [comment]
I like that. Thanks for sharing.

If we want tunings for common software packages, we need a facility to accept them.
If people post FPs, it's hard to update the existing rules with exceptions. If we provide separate files and a facility to host those exclusion rules, it's fairly easy for us and even the reporter to provide exclusion rules in a predefined format.

Given that reasoning and thinking some more, I am changing my mind and favor the introduction of such a mechanism before RC2. If we postpone it to 3.1.0, we might lose momentum or credibility.

But let's not use 960-979. That's a break in the namespace concept. Let's think big and let's not limit ourselves to 19 applications. And let's break free of the 900K range for this.

How about this?
```
REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf.example
REQUEST-901-INITIALIZATION.conf
REQUEST-902.9001-DRUPAL-EXCLUSION-RULES.conf
REQUEST-902.9002-WORDPRESS-EXCLUSION-RULES.conf
REQUEST-905-COMMON-EXCEPTIONS.conf
REQUEST-910-IP-REPUTATION.conf
```

The rules in question would then be hosted in the 9M range. 9,001,000 for Drupal, 9,002,000 for Wordpress, etc.  We are afraid of rule id misunderstandings, we would take another of the ranges >1M. If we pick 9M, we can talk of "900K" or "9M" rule.

I think this saves the namespace concept and gives us plenty of room for all future developments.

**Comment by lifeforms** [comment]
@dune73 Yeah that makes more sense. Logically they should be placed before the other rules.

**Comment by dune73** [comment]
To be honest I would not mind swapping 900 and 901. That way we do initialization first and run the exclusion rules afterwards. Unless of course, there are reasons to disable rules in the initialisation. Then this would probably be a bad idea.

**Comment by csanders-git** [comment]
one issue to think about here is using ctl: is expensive versus secruleupdattarget by id

**Comment by dune73** [comment]
Yes. 

But we can't make `SecRuleUpdateTargetBy(Id|Msg|Tag)` switchable. Or can we?

**Comment by lifeforms** [comment]
We could put it in a `<LocationMatch>` although that would tie us to Apache - but we couldn't be very precise. The ctls will only happen in very small percentage of requests though - in this example the URI must match, ARGS:ajax_html_ids[] must be there, etc. before it's ran, which will probably be in an extreme minority of requests.

**Comment by dune73** [comment]
God forbid a `<LocationMatch>`!

Is not part of the cost of the `ctl` statements the fact that there are unnecessary matching operations which will fail in 99% of the requests?

As for your speedup idea: I support that. Even more so as it makes the rule more readable thanks to the key component being checked first.

I thought about `SecRuleUpdateTargetBy(Id|Msg|Tag)`. Maybe we could work with apache mod_define and then do IF-expressions in the apache config and thus includes depending on these values. This would mean to introduce a new file sourcing mechanism though and I am not a fan.

And finally, before I forget it. Can we make the URI check depend on a custom filename prefix in a variable?

**Comment by lifeforms** [comment]
@dune73 I had thought about an uri prefix, but I want the CRS to remain ergonomic for hosters running Drupal in every possible path server wide. So I wouldn't want to require people typing stuff  like `tx.crs_drupal_path=/.*` to ensure matches on all possible paths. But we could accommodate that by having this path as the default.

If we do path checks, we should circumvent this by setting smart defaults (like assume `/.*` if the user doesn't set the path). If we do a hard check on the full path,  then we would need to use macro expansions in every REQUEST_FILENAME/URI match, and be forced to use regexps instead of faster `@endsWith`, but that's acceptable.

Although, at the other hand, having the path prefix available in a variable, and having a policy to always match using the full path, would help us prevent some possible false negatives. Exclusions can become more precise. For instance, the above would exclude `/drupal/file/ajax/submitted` (which we want) but it would also exclude ``/drupal/some/more/subdirs/file/ajax/submitted` (which might open up a false negative). We could prevent it by always doing a full path regexp match. We would then write our rules always like:

```
SecRule REQUEST_URI "^%{tx.crs_drupal_path}/file/ajax/(submitted|upload)/" \
```

If we don't do this, we can still accommodate people specifying a path if they do something like this:

```
SecRule REQUEST_URI "@beginsWith /drupal/" \
     "id:900860,phase:1,nolog,pass,t:none,setvar:tx.crs_tuning_drupal=1"
```

Only the inaccuracy of the exclusion path remains.

Reconsidering it, I actually agree we should do a path match, as long as we set it to *any* path by default.

**Comment by dune73** [comment]
Back on this topic.

First a remark on @lifeform's last comment: We can not do macro expansion with `@rx`. It does not support that. We would need to work with `@beginsWith`.

@lifeforms has a point with the plan to be as granular as possible. But if we aim for a lot of supported apps, then performance becomes a liability (with skips to be executed for every request outside of the said ap. That is 99% of requests on many, many servers. So @csanders-git does have a point when calling for `SecRuleUpdateTargetById` and friends.

So I sat down and tried to come up with a conditional example which lets us leverage `SecRuleUpdateTargetById` conditionally. Based on issue #527, I came up with the following:

```

# Enable the following applications if installed
Define PICASA_ENABLED
#Define FOO_ENABLED
#Define BAR_ENABLED

<IfDefine PICASA_ENABLED>
    Include /.../application-specific-exclusion-rules-after-crs/9001000-picasa.conf
</IfDefine>

```

`9001000-picasa.conf` would then be 

```
SecRuleUpdateTargetById 9210 !FILES_NAMES
```

Obviously, this is apache syntax and we would need three different rule files performing the conditional include for Apache, nginX and IIS. Let alone custom implementation.

On top of it, the construct above would disable the rule for the complete service - not only Picasa which is no longer granular, but very wide.

So maybe the approach to exclude rules via ctl statements is better. But we really need to avoid dozens of includes. But actually, that only applies in the long run. As long as it's only 4-5 applications we support, the skipping mechanism based on `TX.crs_tuning_xxx` is good enough.

**Comment by dune73** [comment]
@lifeforms : I was wrong about macro expansion with @rx. It is supported. Recently read that it was not, did not check and now I do not longer remember where I read it.

**Comment by dune73** [comment]
Let's close this in favor of #579.

**Comment by lifeforms** [comment]
Decided on IRC:

- Yes, do app exclusions
- Start with WordPress, Drupal
- Make an on/off switch per app
- Don't use global flag
- Don't filter on paths (either using regexp or beginsWith) - regexp would be complex to configure and probably an advanced feature; strings don't offer much advantage over the user using SecRule to selectively enable exclusions if wanted.
